File: 算法/代码随想录算法PDF/00调试源代码/139.单词拆分.cpp
/*
 * @lc app=leetcode.cn id=139 lang=cpp
 *
 * [139] 单词拆分
 */

// @lc code=start
class Solution {
public:
    bool wordBreak(string s, vector<string>& wordDict) {
        vector<bool> dp(s.size() + 1, false);
        dp[0] = true;
        for (int i = 1; i <= s.size(); i++) {
            for (string word : wordDict) {
                if (i >= word.size()) {
                    dp[i] = dp[i] || (dp[i - word.size()] && word == s.substr(i - word.size(), word.size()));
                }
            }
        }
        return dp[s.size()];
    }
};
// @lc code=end

File: 算法/代码随想录算法PDF/00调试源代码/139.单词拆分.py
#
# @lc app=leetcode.cn id=139 lang=python3
#
# [139] 单词拆分
#

# @lc code=start
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        dp = [False] * (len(s) + 1)
        dp[0] = True
        for i in range(1, len(s) + 1):
            for word in wordDict:
                if i >= len(word):
                    dp[i] = dp[i] or (dp[i - len(word)] and word == s[i - len(word) : i])
        return dp[-1]
# @lc code=end

